### PR TITLE
Sphynx Cat - Kafka DLQs Epic (GSI-827)

### PR DESCRIPTION
**Pull request by TheByronHimes**
Adds the tech spec for the proposed Kafka Dead Letter Queues epic (Sphynx Cat), which includes an ADR, some hexkit changes, and then the resulting implementation rollout across services.

I would appreciate any feedback you feel like offering, especially with regard to the implementation concept itself. I hope the diagrams and hexkit class descriptions offer enough to go on. I have an ADR posted [here](https://github.com/ghga-de/adrs/pull/13)

### CODE REVIEW

**Review by lkuchenb — CHANGES_REQUESTED** [review]
Thank you, great work!

I have no concerns with the proposal at this point, looking forward to the feedback from others.

I would request to extend the epic to include writing an asynchronous service communication architecture concept over in the internal docs, maybe with the following structure:

* Brief intro that we are using async comm between services based on Kafka
* Usage of compacted topics, what does that mean for the types of messages that we publish, which rules do we have to adhere to
* Outbox pattern, reasoning and further implications / rules due to that
* DQL concept

For the DLQ part you can basically copy & paste this markdown and update whatever has changed conceptually during implementation. For outbox and compacted topics there's probably also recyclable material, either way it would be great to document this while it's still fresh in your head :)

File: 56-sphynx-cat/technical_specification.md [hunk 1]
```diff
@@ -0,0 +1,151 @@
...
+One of the most common mechanisms for achieving this capability is called a
+Dead Letter Queue (DLQ). In a DLQ, failed events are sent to a separate location. In the
+context of Kafka, this is usually another topic, but it could also be a database
+collection or any other location that effectively preserves the event while preventing
+further processing. Kafka does not come with DLQ support out-of-the-box unless used with
+Kafka Connect. We are not using Kafka Connect and use AIOKafka's python producers and
+consumers with our own library instead, meaning we can't take advantage of Kafka Connect
+without considerable work.
+
+> It's important to note that using a dead letter queue *does not* remove the need for
```
[the remark this answers is not in the document]
Comment (TheByronHimes) [reply 56-sphynx-cat/technical_specification.md:36]: Nice! I didn't know about that feature.

File: 56-sphynx-cat/technical_specification.md [hunk 2]
```diff
@@ -0,0 +1,151 @@
...
+### Implementation of DLQ Providers in `hexkit`
+
+The work will involve the creation of a class that combines the Kafka Subscriber and the
+Kafka Publisher, which will be used in two ways.
+
+
+**`KafkaRetrySubscriber`** *(name subject to change)*  
+A specialized event consumer that, upon catching an unhandled error during event
+consumption, does the following:
+1. Retries the event until the configured number of retries is exhausted (using
```
Comment (Cito) [inline 56-sphynx-cat/technical_specification.md:85]:
```suggestion
1. Retries handling the event until the configured number of retries is exhausted (using
```

File: 56-sphynx-cat/technical_specification.md [hunk 3]
```diff
@@ -0,0 +1,151 @@
...
+exponential backoff).
+   - This can reduce the likelihood of transient errors populating the DLQ.
+1. Checks auto-ignore rules to determine if the event should be dropped or published
+   - This can be implemented later when it is more clear how best to do this.
+2. Publishes the event to a service-specific DLQ topic.
+   1. The original topic should be added as a field appended to the payload.
```
Comment (Cito) [inline 56-sphynx-cat/technical_specification.md:91]:
```suggestion
   - The original topic should be added as a field appended to the payload.
```

File: 56-sphynx-cat/technical_specification.md [hunk 4]
```diff
@@ -0,0 +1,169 @@
...
+3. Commits the topic offsets to avoid infinitely reprocessing the failed event.
+4. Upon consuming an event from the service-specific retry topic, extracts/removes the
```
Comment (mephenor) [inline 56-sphynx-cat/technical_specification.md:93]: Numbering is weird in markdown, though it's correctly rendered fully in sequence

File: 56-sphynx-cat/technical_specification.md [hunk 5]
```diff
@@ -0,0 +1,169 @@
...
+This class is potentially not required if requeueing is done through a 3rd party tool
+like Kafka UI. However, if requeueing is done through either the different services or
+an in-house dedicated DLQ service, then this type of consumer will be helpful.
+
+## Other: 
+
+### Potential Problems:
+
+> [!NOTE]
+> Note: The following list is by no means exhaustive:
```
Comment (mephenor) [inline 56-sphynx-cat/technical_specification.md:114]:
```suggestion
> The following list is by no means exhaustive: 
```
This is superfluous with the [!NOTE] directly above

**Review by mephenor — DISMISSED** [review]
I +1 what Leon already said, including the request for an asynchronous service communication architecture concept in the internal docs.

Apart from that, only some slight nitpicking remains.

File: 56-sphynx-cat/technical_specification.md (hunk 5 above)
Comment (TheByronHimes) [reply 56-sphynx-cat/technical_specification.md:114]: Good eye, sniper

**Review by Cito — APPROVED** [review]

**Review by Cito — APPROVED** [review]

**Review by mephenor — DISMISSED** [review]

**Review by lkuchenb — APPROVED** [review]

**Pull request merged**